### PR TITLE
Add support for explicit nondet in java code

### PR DESCRIPTION
**Pull request by reuk**
This PR supersedes #607, implementing the same functionality. We block the loading and conversion of the nondet Java library functions, which means that the functions returning primitives use automatic nondet stubbing. For nondet objects, instead of relying on automatic nondet stubbing, we add a dedicated modification pass over the goto program. This pass manually calls `gen_nondet_init` to generate goto-program instructions to nondet-initialise nested structures.

### CODE REVIEW

**Comment by peterschrammel** [comment]
https://github.com/diffblue/cbmc/pull/691/commits/7c52151af7b2b52ee82f4b78ad0d01ecfdb2cf3e must go into a separate PR

**Comment by peterschrammel** [comment]
The reordering of header files in https://github.com/diffblue/cbmc/pull/691/commits/bd366ce52fc03e7f2153a5e83f76e754b4e2f27e must go into a separate PR

File: src/cbmc/cbmc_parse_options.cpp [hunk 1]
```diff
@@ -907,6 +907,8 @@ bool cbmc_parse_optionst::process_goto_program(
     // Similar removal of RTTI inspection:
     remove_instanceof(symbol_table, goto_functions);
     // Similar removal of java nondet statements:
+    // TODO Should really get this from java_bytecode_language somehow, but we
+    // don't have an instance of that here.
```
Comment (peterschrammel) [inline src/cbmc/cbmc_parse_options.cpp]: at least MAX_NONDET_ARRAY_LENGTH_DEFAULT should be used

Comment (peterschrammel) [inline src/cbmc/cbmc_parse_options.cpp]: A fix for this could be to pass the optionst reference through language_ui so that the options parsed there are also available here. I don't like that all these options are stored in separate variables in java_bytecode_languaget anyway.

File: src/goto-programs/remove_java_nondet.cpp [hunk 2]
```diff
@@ -0,0 +1,389 @@
...
+}
+
+/*******************************************************************\
+
+Function: get_nondet_instruction_info
+
+  Inputs:
+    instr: A goto program instruction.
+
+ Outputs: Whether the instruction is a function with a nondet signautre, and
```
Comment (martin-cs) [inline src/goto-programs/remove_java_nondet.cpp]: s/signautre/signature/

File: src/goto-programs/remove_java_nondet.cpp [hunk 3]
```diff
@@ -0,0 +1,389 @@
...
+  const auto &code=to_code(instr->code);
+  if(code.get(ID_statement)!=ID_function_call)
+  {
+    return {false, false};
+  }
+  const auto &function_call=to_code_function_call(code);
+  return is_nondet_returning_object(function_call);
+}
+
+// TODO Pascal suggested moving the following function to util, but I'm not
```
Comment (martin-cs) [inline src/goto-programs/remove_java_nondet.cpp]: Resolve TODO

File: src/goto-programs/remove_java_nondet.cpp [hunk 4]
```diff
@@ -0,0 +1,389 @@
...
+// sure.  If goto_programt has public data members, this implies that any state
+// that these data members takes is valid for the goto_program as a whole,
+// which is obviously not the case - if I insert a node into
+// goto_programt::instructions without updating the iterator positions, this
+// has a good chance of invalidating the program.  It should be *impossible* to
+// place an object (with the strong exception safety guarantee) into an invalid
```
Comment (martin-cs) [inline src/goto-programs/remove_java_nondet.cpp]: Yes; this is a known issue and is documented elsewhere.  Fixing this is an entirely different task.  There are not only structural but semantic issues here.  When I / someone in QA has a chance, we need a "check well formed" method for goto_functiont that at least documents some of the invariants that must hold on the structure.

File: src/goto-programs/remove_java_nondet.cpp [hunk 5]
```diff
@@ -0,0 +1,389 @@
...
+    {
+      const auto new_target=target_map.find(t);
+      assert(new_target!=std::end(target_map));
+      t=new_target->second;
+    }
+  }
+
+  return pos;
+}
+
```
Comment (martin-cs) [inline src/goto-programs/remove_java_nondet.cpp]: I think goto_programt::insert_before() then goto_programt::update() will do what you want.

File: src/goto-programs/remove_java_nondet.cpp [hunk 6]
```diff
@@ -0,0 +1,389 @@
...
+ Purpose: Return whether the expression is a symbol with the specified
+          identifier.
+
+\*******************************************************************/
+
+static bool is_symbol_with_id(const exprt& expr, const dstringt& identifier)
+{
+  return expr.id()==ID_symbol && expr.get(ID_identifier)==identifier;
+}
```
Comment (martin-cs) [inline src/goto-programs/remove_java_nondet.cpp]: As a general rule, don't use get() directly; convert to the appropriate exprt type and use the accessor methods.

File: src/goto-programs/remove_java_nondet.cpp [hunk 7]
```diff
@@ -0,0 +1,389 @@
...
+  const auto &code_assign=to_code_assign(matching_assignment->code);
+  const auto nondet_var=code_assign.lhs();
+  const auto source_loc=instr->source_location;
+
+  // Erase from the nondet function call to the assignment
+  const auto after_matching_assignment=std::next(matching_assignment);
+  assert(after_matching_assignment!=end);
+  const auto after_erased=
+    instructions.erase(instr, after_matching_assignment);
+
```
Comment (martin-cs) [inline src/goto-programs/remove_java_nondet.cpp]: Hacking the internals of a goto_program directly is not a good idea; use the methods it provides.

File: src/goto-programs/remove_java_nondet.cpp [hunk 8]
```diff
@@ -0,0 +1,389 @@
...
+{
+  // If not an assignment, return false
+  if(!instr.is_assign())
+  {
+    return false;
+  }
+  const auto &rhs=to_code_assign(instr.code).rhs();
+  return is_symbol_with_id(rhs, identifier) ||
+         is_typecast_with_id(rhs, identifier);
+}
```
Comment (martin-cs) [inline src/goto-programs/remove_java_nondet.cpp]: As this is the only place these two functions are used, why not put the code here directly?

File: src/goto-programs/remove_java_nondet.cpp [hunk 9]
```diff
@@ -0,0 +1,389 @@
...
+  const size_t max_nondet_array_length,
+  goto_programt &goto_program)
+{
+  // Check each instruction.
+  // `process_target` may modify the list in place, and returns the next
+  // iterator to look at.
+  for(auto instruction_iterator=std::begin(goto_program.instructions),
+        end=std::end(goto_program.instructions);
+      instruction_iterator!=end;
+      instruction_iterator=process_target(
```
Comment (martin-cs) [inline src/goto-programs/remove_java_nondet.cpp]: use forall_goto_program_instructions

File: src/goto-programs/remove_java_nondet.cpp [hunk 10]
```diff
@@ -0,0 +1,389 @@
...
+        message_handler,
+        symbol_table,
+        max_nondet_array_length,
+        goto_program.instructions,
+        instruction_iterator))
+  {
+    // Loop body deliberately empty.
+  }
```
Comment (martin-cs) [inline src/goto-programs/remove_java_nondet.cpp]: Separate out the iteration and the action on the instruction; merging the two is unnecessarily complex.

File: src/goto-programs/remove_java_nondet.cpp [hunk 11]
```diff
@@ -0,0 +1,389 @@
...
+{
+#ifdef JAVA_NONDET_DEBUG
+  std::cerr
+    << "replacing nondet statements using max array size of: "
+    << max_nondet_array_length
+    << '\n';
+#endif
+
+  for(auto &f : goto_functions.function_map)
+  {
```
Comment (martin-cs) [inline src/goto-programs/remove_java_nondet.cpp]: Use forall_goto_functions

File: src/goto-programs/remove_java_nondet.h [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+
+ Purpose: Modify a goto_functionst to replace 'nondet' library functions with
+          CBMC-native nondet expressions.
+
+\*******************************************************************/
+
+void remove_java_nondet(
+  message_handlert &message_handler,
+  symbol_tablet &symbol_table,
+  size_t max_nondet_array_length,
```
Comment (martin-cs) [inline src/goto-programs/remove_java_nondet.h]: Why are we fixing array length to a constant at this stage?

File: src/java_bytecode/java_bytecode_language.cpp [hunk 13]
```diff
@@ -40,26 +41,57 @@ Function: java_bytecode_languaget::get_language_options
...
+    disable_runtime_checks=cmd.isset("disable-runtime-check"));
+
+  options.set_option(
+    "java-assume-inputs-non-null",
+    assume_inputs_non_null=cmd.isset("java-assume-inputs-non-null"));
+
+  options.set_option(
+    "refine-strings",
+    string_refinement_enabled=cmd.isset("refine-strings"));
+
```
Comment (martin-cs) [inline src/java_bytecode/java_bytecode_language.cpp]: Option parsing from cmdlinet to optiont should be done with the rest of the parse_option code (in the application directory), not in the front-end.

File: src/java_bytecode/java_bytecode_language.h [hunk 14]
```diff
@@ -100,8 +101,8 @@ class java_bytecode_languaget:public languaget
                                     //  - checkcast / instanceof
                                     //  - array bounds check
                                     //  - array size for newarray
-  size_t max_nondet_array_length;   // maximal length for non-det array creation
-  size_t max_user_array_length;     // max size for user code created arrays
+  unsigned max_nondet_array_length; // maximal length for non-det array creation
+  unsigned max_user_array_length;   // max size for user code created arrays
   lazy_methodst lazy_methods;
```
Comment (martin-cs) [inline src/java_bytecode/java_bytecode_language.h]: Why has this been changed?  I can see the case for uint32_t but not unsigned.

File: src/java_bytecode/library/Makefile [hunk 15]
```diff
@@ -0,0 +1,26 @@
+all: org.cprover.jar
```
Comment (martin-cs) [inline src/java_bytecode/library/Makefile]: 1. setting up the java library should be a separate PR.
2. isn't someone else working on modelling the Java standard library?

File: src/java_bytecode/library/src/org/cprover/CProver.java [hunk 16]
```diff
@@ -3,9 +3,124 @@
...
+
+  public static void assume(boolean condition)
   {
     if(enableAssume)
-      throw new RuntimeException("Cannot execute program with CProver.assume()");
-  }  
+    {
+      throw new RuntimeException(
+          "Cannot execute program with CProver.assume()");
+    }
```
Comment (martin-cs) [inline src/java_bytecode/library/src/org/cprover/CProver.java]: Can I suggest first checking if the assume is true and only throwing an assertion if it is false, that will make it safe to run traces generated taking the assume into account.

**Comment by martin-cs** [comment]
The approach used is quite different to how nondet is used in C and C++.  Can we have a brief discussion of what the design goals are here?  It's possible you are trying to do something different which needs a different design but if not it would be better to be consistent with what is done in other front-ends.

File: regression/cbmc-java/Makefile [hunk 17]
```diff
@@ -18,3 +18,17 @@ show:
 			vim -o "$$dir/*.java" "$$dir/*.out"; \
 		fi; \
 	done;
+
+
+%.class: %.java ../../src/org.cprover.jar
```
Comment (peterschrammel) [inline regression/cbmc-java/Makefile]: We currently commit the class files 
I'm generally in favour of generating them instead of putting them into the repo. This would allow us to test different jdk versions. However, this causes lots of other complications, such as some tests that target specific jdk versions only or tests that requre bytecodes that we don't know how to generate with any current Java compiler. Hence, the need for class files in the repo won't disappear completely. Anyways, this would concern all Java regression tests and can't just be done for a single directory.

Comment (martin-cs) [inline regression/cbmc-java/Makefile]: OK; I'll leave that for another PR then.

File: src/cbmc/cbmc_parse_options.cpp [hunk 18]
```diff
@@ -74,7 +75,7 @@ Function: cbmc_parse_optionst::cbmc_parse_optionst
 cbmc_parse_optionst::cbmc_parse_optionst(int argc, const char **argv):
   parse_options_baset(CBMC_OPTIONS, argc, argv),
   xml_interfacet(cmdline),
-  language_uit(cmdline, ui_message_handler),
+  language_uit(cmdline, ui_message_handler, std::make_shared<optionst>()),
```
Comment (peterschrammel) [inline src/cbmc/cbmc_parse_options.cpp]: The policy is to always use the least appropriate means to solve a problem. I think this is not the case here. I don't see why shared ownership is required.

The problem that Reuben tries to solve here is to avoid duplicate parsing of options in language_Ui and *_parse_options, which is particularly annoying when default values are used when an option is not given. Therefore the idea is to use the options object also in language_Ui.

Yes, that's a major refactoring that requires a separate PR to introduce it.

Comment (reuk) [inline src/cbmc/cbmc_parse_options.cpp]: The problem is that we need to use the java max nondet array length, which is set from the command line. Rather than parsing the cmdline options twice, we should allow the java frontend to parse out the options, and then query these. Therefore language_uit needs to store a reference to an optionst which will store the parsed options. From @peterschrammel's feedback above:

> A fix for this could be to pass the optionst reference through language_ui so that the options parsed there are also available here. I don't like that all these options are stored in separate variables in java_bytecode_languaget anyway.

Re: shared_ptr, I'm just using this so that there's no possibility that we try to deref a null ptr/ref to an optionst. Another possibility is for language_uit to have an optionst as a direct data member. I'm reluctant to store a ref or raw-ptr to an optionst because it's too easy for the pointed-to object to go out-of-scope before the language_uit instance.

**Comment by peterschrammel** [comment]
Language_Ui cannot have options as a direct member,  see goto-diff. Parse_options could, however.

File: src/cbmc/cbmc_parse_options.cpp (hunk 18 above)
Comment (reuk) [inline src/cbmc/cbmc_parse_options.cpp]: Re: shared_ptr, I'm just using this so that there's no possibility that we try to deref a null ptr/ref to an optionst. Another possibility is for language_uit to have an optionst as a direct data member. I'm reluctant to store a ref or raw-ptr to an optionst because it's too easy for the pointed-to object to go out-of-scope before the language_uit instance.

Comment (peterschrammel) [inline src/cbmc/cbmc_parse_options.cpp]: Oups, commas are important: least, appropriate :D

File: src/goto-programs/remove_java_nondet.h (hunk 12 above)
Comment (reuk) [inline src/goto-programs/remove_java_nondet.h]: We are calling `gen_nondet_init` to generate the nondet initialisation code, which takes a max array size.

File: src/goto-programs/remove_java_nondet.cpp (hunk 10 above)
Comment (reuk) [inline src/goto-programs/remove_java_nondet.cpp]: The action may delete the current instruction and insert a range of new instructions. It is not sufficient to simply iterate over the instructions (using `forall_goto_program_instructions`) because the list iterator may no longer point to a valid node in the list by the end of the loop body. Besides this, after inserting instructions, we do not need to check each of the inserted instructions. Instead, we should resume checking from after the inserted range. Therefore I believe the current method is justified.

File: src/goto-programs/remove_java_nondet.cpp (hunk 11 above)
Comment (reuk) [inline src/goto-programs/remove_java_nondet.cpp]: What's the advantage vs the range-for? forall_goto_functions hides whether the iteration is mutable or const, computes the end iterator on each iteration, and post-increments rather than pre-increments the iterator, which may lead to an additional copy per iteration. Range-for seems strictly better in terms of readability and performance.

File: src/goto-programs/remove_java_nondet.cpp (hunk 7 above)
Comment (reuk) [inline src/goto-programs/remove_java_nondet.cpp]: goto_program does not appear to provide a range erase

File: src/java_bytecode/java_bytecode_language.h (hunk 14 above)
Comment (reuk) [inline src/java_bytecode/java_bytecode_language.h]: This was changed so that the call to `optionst::set_option` would not be ambiguous.

File: src/java_bytecode/library/src/org/cprover/CProver.java (hunk 16 above)
Comment (reuk) [inline src/java_bytecode/library/src/org/cprover/CProver.java]: Assume isn't mine, I just fixed the formatting.

File: src/goto-programs/remove_java_nondet.cpp (hunk 8 above)
Comment (reuk) [inline src/goto-programs/remove_java_nondet.cpp]: I think this is more readable and self-documenting.

File: src/java_bytecode/library/src/org/cprover/CProver.java (hunk 16 above)
Comment (reuk) [inline src/java_bytecode/library/src/org/cprover/CProver.java]: Should this change be made in this PR, given that assume was added in a previous PR and has not been modified (other than formatting) in this PR?

**Comment by reuk** [comment]
@martin-cs I'd be happy to discuss design goals.

I've fixed-up or responded to each of the comments, other than those to do with cmdline parsing. What's the best course of action here? I can't think of a good solution which doesn't introduce the possibility of dangling references - I think adding an optionst member field to parse_options_base means holding a reference in language_uit, which makes me uncomfortable

File: src/goto-programs/remove_java_nondet.cpp (hunk 9 above)
Comment (reuk) [inline src/goto-programs/remove_java_nondet.cpp]: See my response on line 364

**Comment by reuk** [comment]
After discussing with @peterschrammel I have submitted an issue (#741) against the current command-line parsing structure, which can separately address duplication of option-parsing code. I have rolled-back the changes to command-line parsing on this PR.

**Comment by martin-cs** [comment]
On Thu, 2017-03-30 at 10:24 -0700, Reuben Thomas wrote:
> @martin-cs I'd be happy to discuss design goals.

Why don't we discuss the whole PR / review as it'll be more efficient
than doing everything via github.

File: src/goto-programs/remove_java_nondet.cpp (hunk 7 above)
Comment (peterschrammel) [inline src/goto-programs/remove_java_nondet.cpp]: Then we should add it to goto_program. Do this in a separate PR.

File: src/goto-programs/remove_java_nondet.cpp [hunk 19]
```diff
@@ -0,0 +1,330 @@
...
+#include "goto-programs/goto_model.h"
+
+#include "java_bytecode/java_object_factory.h"
+
+#include "util/irep_ids.h"
+
+#include <algorithm>
+#include <regex>
+
+/*******************************************************************\
```
Comment (peterschrammel) [inline src/goto-programs/remove_java_nondet.cpp]: All this should be wrapped into a class as done for other remove_* passes.

File: src/goto-programs/remove_java_nondet.cpp (hunk 9 above)
Comment (peterschrammel) [inline src/goto-programs/remove_java_nondet.cpp]: We currently use forall_goto_program_instructions throughout the code base. Using something else is a global change to be applied to the whole code base.

File: src/java_bytecode/library/Makefile (hunk 15 above)
Comment (peterschrammel) [inline src/java_bytecode/library/Makefile]: 1. Fine with me if it is in separate commits.
2. Reuben is working on  the Java library. This PR is a prerequisite to enable nondet+assume patterns.

File: src/java_bytecode/library/src/org/cprover/CProver.java (hunk 16 above)
Comment (peterschrammel) [inline src/java_bytecode/library/src/org/cprover/CProver.java]: Do it in a separate commit

File: src/goto-programs/remove_java_nondet.cpp [hunk 20]
```diff
@@ -0,0 +1,330 @@
...
+  const auto source_loc=instr->source_location;
+
+  // Erase from the nondet function call to the assignment
+  const auto after_matching_assignment=std::next(matching_assignment);
+  assert(after_matching_assignment!=end);
+  const auto after_erased=
+    instructions.erase(instr, after_matching_assignment);
+
+  // Generate nondet init code
+  code_blockt init_code;
```
Comment (peterschrammel) [inline src/goto-programs/remove_java_nondet.cpp]: I think this should be done as follows:
1. remove here the nondet_*() calls and replace them by return_value = NONDET; for objects the classcast must be considered to get the right return type.
2. in a separate, language-independent goto transformation pass, e.g. remove_nondet(), elaborate the NONDET of complex types by appropriate initialisation code. This has the advantage that the nondet-initialisation code that is currently present in *_entry_point can go away and is uniformly handled by this new pass. Moreover, it's not language-dependent anymore and more easily maintainable.

File: src/goto-programs/remove_java_nondet.cpp (hunk 19 above)
Comment (reuk) [inline src/goto-programs/remove_java_nondet.cpp]: This change harms readability, and (counter-intuitively) *increases* the scope of the member variables. Consider:

```c++
class nondet_removert final
{
public:
  nondet_removert(...): /* set data members */ {}
  //  all the methods in the module
private:
  message_handlert &message_handler;
  symbol_tablet &symbol_table;
  size_t max_nondet_array_length;
  goto_functionst &goto_functions;
};
```

Now, all the methods in `nondet_removert` may access and modify any of the data members. Currently, each function in the module may only read the parameters that it is passed, which means that when reading the code the reader only has to keep the function parameters and local variables in their head. Making the functions into member functions means that the reader has to keep the entire state of the object in their head. Additionally, member functions would mean that each function may *modify* the data members, which creates the possibility of undesired side-effects.

This change is essentially asking for a set of unrelated variables to be shared amongst a set of unrelated functions. This is bad, in the same way that global variables are bad (albeit at a smaller scale).

Also, the conceptual need for a class is not clear. Classes serve three main purposes in C++:
- resource management via RAII
- polymorphism and customisation via virtual methods
- encapsulation/maintaining invariants over data via data hiding and public interfaces

A class would not take on any of these roles in this context, which makes me question its necessity.

Finally, this change increases the sheer quantity of code to maintain (class declaration, constructor, documentation block).

**Comment by reuk** [comment]
Now depends on #760.

**Comment by reuk** [comment]
I believe this is ready for review again.

File: src/goto-programs/replace_java_nondet.cpp [hunk 21]
```diff
@@ -0,0 +1,319 @@
...
+    identifier: Some identifier.
+
+ Outputs: True if the expression is a symbol with the specified identifier.
+
+ Purpose: Return whether the expression is a symbol with the specified
+          identifier.
+
+\*******************************************************************/
+
+static bool is_symbol_with_id(const exprt& expr, const dstringt& identifier)
```
Comment (peterschrammel) [inline src/goto-programs/replace_java_nondet.cpp]: use `irep_idt`

File: src/goto-programs/replace_java_nondet.cpp [hunk 22]
```diff
@@ -0,0 +1,319 @@
...
+
+ Outputs: True if the expression is a typecast with one operand, and the
+          typecast's identifier matches the specified identifier.
+
+ Purpose: Return whether the expression is a typecast with the specified
+          identifier.
+
+\*******************************************************************/
+
+static bool is_typecast_with_id(const exprt& expr, const dstringt& identifier)
```
Comment (peterschrammel) [inline src/goto-programs/replace_java_nondet.cpp]: use `irep_idt` (several occurrences)

File: src/goto-programs/replace_java_nondet.cpp [hunk 23]
```diff
@@ -0,0 +1,319 @@
...
+        end=goto_program.instructions.cend();
+      instruction_iterator!=end;)
+  {
+    instruction_iterator=check_and_replace_target(
+      goto_program,
+      instruction_iterator);
+  }
+}
+
+
```
Comment (peterschrammel) [inline src/goto-programs/replace_java_nondet.cpp]: function comment block missing

File: src/java_bytecode/library/src/org/cprover/CProver.java (hunk 16 above)
Comment (peterschrammel) [inline src/java_bytecode/library/src/org/cprover/CProver.java]: @reuk, you can add this change to this PR (in a separate commit, of course)

File: src/goto-programs/convert_nondet.cpp [hunk 24]
```diff
@@ -0,0 +1,164 @@
...
+
+\*******************************************************************/
+
+static void convert_nondet(
+  goto_programt &goto_program,
+  symbol_tablet &symbol_table,
+  message_handlert &message_handler,
+  size_t max_nondet_array_length)
+{
+  for(auto instruction_iterator=goto_program.instructions.cbegin(),
```
Comment (reuk) [inline src/goto-programs/convert_nondet.cpp]: `forall_goto_program_instructions` can't be used here because the loop body should not visit all goto program instructions.

File: src/goto-programs/replace_java_nondet.cpp (hunk 23 above)
Comment (reuk) [inline src/goto-programs/replace_java_nondet.cpp]: The comment block is in the header.

**Comment by reuk** [comment]
I've fixed or responded to the above items.
